src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -780,6 +797,64 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
                 parts_with_ranges[part_index] = std::move(ranges);
         };
 
+        auto check_and_create_new_key_condition_from_pk_ranges = [&]()
+        {
+            size_t range_count = 0;
+            size_t first_part_for_next_pass = UINT64_MAX;
+
+            if (!is_select_query_with_final || !was_any_skip_index_useful || !settings[Setting::use_skip_indexes_if_final_exact_mode])
+                return std::make_tuple(false, static_cast<size_t>(0), static_cast<KeyConditionPtr>(nullptr));
+
+            const auto & primary_key = metadata_snapshot->getPrimaryKey();
+            KeyConditionPtr kc = make_shared<KeyCondition>(nullptr, context, primary_key.column_names, primary_key.expression);
+            for (size_t part_index = 0; part_index < parts.size(); ++part_index)
+            {
+                auto & part = parts[part_index];
+                const auto & index = part->getIndex();
+
+                if (index->empty())
+                    continue;
+
+                auto ranges = parts_with_ranges[part_index];
+                auto index_columns = std::make_shared<ColumnsWithTypeAndName>();
+                for (size_t i = 0; i < index->size(); i++)
+                    index_columns->emplace_back(index->at(i), primary_key.data_types[i], primary_key.column_names[i]);
+
+                /// Need to build : (((pk1 BETWEEN 5 AND 10) AND (pk2 BETWEEN 100 AND 110)) OR ((pk1 BETWEEN 1000 AND 1005) AND (pk2 BETWEEN 2000 AND 2100)) OR ... )
+                for (auto range : ranges.ranges)
+                {
+                    std::vector<FieldRef> index_left(index->size());
+                    std::vector<FieldRef> index_right(index->size());
+                    for (size_t i = 0; i < index->size();i++)
+                    {
+                        index_left[i] = {index_columns.get(), range.begin, i};
+                        index_right[i] = {index_columns.get(), range.end, i};
+                        kc->addPrimaryKeyRangeCondition(primary_key.column_names[i],
+                                               Range(index_left[i], true, index_right[i], true), (i > 0));
+                    }
+                    range_count++;
+                    if (range_count >= 2)
+                    {
+                          kc->addAnOR();
+                    }
+                    if (first_part_for_next_pass == UINT64_MAX)
+                        first_part_for_next_pass = part_index; /// we only need to check in newer parts
+                }
+                static constexpr size_t ranges_threshold = 1000;
+                if (range_count > ranges_threshold)
```

Comment:
@nickitat Thank you picking up this PR for review! Let me know your overall feedback and if you have any queries. Specifically here, do you think this thresholding is required? I added this protection, but now feel that even if there are more ranges, the slightly additional cost of the "PK search" 2nd pass should be okay. For tables with billions of rows, this limit of 1000 ranges returned by secondary index looks really low.

Reply:
i think the main challenge here is to make it efficient for wider set of tables and queries. number of parts could be big (10^3-10^4), number of ranges even bigger by a couple of orders of magnitude. 
we likely need a threshold to abandon the optimisation at some point. but before that we could:
1. try to "compact" PK conditions. e.g. if we have conditions like `in [1; 3]`, `[5; 8]` - we could merge those into a single one - `in [1; 8]`. this way we could maintain number of conditions limited. also possible to decide at some point that we got too far and it won't filter anything and give up on the optimisation.
2. what i was talking about in the issue: we extract non-intersecting parts of PK and don't apply merging algorithm to them in runtime. https://github.com/ClickHouse/ClickHouse/blob/380d55b00ce6a92287dccffa824f20df3dcf0d8f/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L1361-L1375 seems like we also could apply skipping indices to them as if they were parts from ordinary MergeTree tables.